### PR TITLE
gen4 planner: Derived table handling on operators

### PR DESCRIPTION
**Pull request by systay**
## Description
As part of the ongoing effort to improve the Vitess query planner and transition more planner logic to the operator model, this PR focuses on enhancing the handling of derived tables.

Previously, derived tables were only managed effectively if they could be pushed down under a Route. With this update, the query planner is now able to handle situations where derived tables need to be split into multiple operators under the Route. This improvement enables the queries sent to MySQL to more closely resemble the user's original query, retaining the derived table in the query being sent down.

By maintaining the derived table in the query, the result is easier for users to understand, as it better reflects their initial input. This enhancement aligns with the overall goal of simplifying the query planning process, allowing developers to better comprehend the optimization steps and create more efficient and optimized queries for their applications.

## Related Issue(s)
Tracking issue #11626

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/sqlparser/tracked_buffer.go [hunk 1]
```diff
@@ -37,10 +37,19 @@ type TrackedBuffer struct {
...
-	escape        bool
 	fast          bool
+
+	escape escapeType
 }
 
+type escapeType int
+
+const (
+	EscapeKeywords escapeType = iota
```
Comment (systay) [inline go/vt/sqlparser/tracked_buffer.go:48]: should not be public

File: go/vt/sqlparser/tracked_buffer.go [hunk 2]
```diff
@@ -37,10 +37,19 @@ type TrackedBuffer struct {
...
+	escapeAllIdentifiers
+	EscapeNoIdentifiers
```
Comment (frouioui) [inline go/vt/sqlparser/tracked_buffer.go:50]: This is only used in `sqlparser`, you can make it private

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 3]
```diff
@@ -195,6 +189,40 @@ func pushDownProjectionInApplyJoin(
 		src.ColumnsAST = append(src.ColumnsAST, col)
 	}
 
+	if p.TableID != nil {
+		derivedTbl, err := ctx.SemTable.TableInfoFor(*p.TableID)
+		if err != nil {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:194]: `pushDownProjectionInApplyJoin` is very chunky, how about we split it in several sub-methods?

File: go/vt/sqlparser/tracked_buffer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/sqlparser/tracked_buffer.go:48]: yes, but it's also the default value, which is what we use when we create the buffer

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go (hunk 3 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/horizon_planning.go:194]: it's just big-boned?

**Review by frouioui — APPROVED** [review]

File: go/vt/sqlparser/tracked_buffer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (frouioui) [reply go/vt/sqlparser/tracked_buffer.go:48]: Got it okay!

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go (hunk 3 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/operators/horizon_planning.go:194]: Yeah a little bit, I don't have strong opinions

File: go/vt/vtgate/semantics/semantic_state.go [hunk 4]
```diff
@@ -138,6 +138,19 @@ func (st *SemTable) CopyDependencies(from, to sqlparser.Expr) {
...
+		return
+	}
+	t, ok := to.(sqlparser.Expr)
+	if !ok {
+		return
+	}
+	st.CopyDependencies(f, t)
+}
+
 // Cloned copies the dependencies from one expression into the other
```
Comment (harshit-gangal) [inline go/vt/vtgate/semantics/semantic_state.go:154]: looks like this is not used

**Pull request merged**